### PR TITLE
19992 rebuild es client on illegal state exception

### LINKED ISSUE #19992: Elasticsearch RestHighLevelClient does not handle RuntimeExceptions
This could be named: dotCMS is unable to recover if we get a
```
java.lang.IllegalStateException: Request cannot be executed; I/O reactor status: STOPPED 
```
For various reasons, RuntimeExceptions get thrown when interacting with Elasticsearch. We do it in our code when searching or putting content to ES or searching content or it can happen outside of dotCMS in the case of memory or networks issues.   I think most people using Elasticsearch expect the `RestHighLevelClient ` code higher up in the stack to catch these exceptions and deal with them appropriately.   The issue is that Elasticsearch's `RestHighLevelClient` and the ES `Request` Objects are built to catch `Exceptions` and not `RuntimeExceptions`.  When a `RunTimeException` is thrown, what happens instead is the `RunTimeException` continues to propagate upwards until it hits the Apache HttpClient and basically destroys the Apache HttpClient connection. When this happens, the HttpClient does not recover and only a server restart will rebuild the HTTPClient - in a perfect world the HttpClient Connection should be rebuilt but this does not seem to happen.

There are 2 things we need to do: 
1. We need to rebuild our `RestHighLevelClient` when we see an `IllegalStateException` `reactor status: STOPPED `
2. We need to wrap all our calls that expect to use the `RestHighLevelClient` in trys that catch `RuntimeExceptions` and rethrow them as `DotElasticsearchException`

### For reference:
elastic/elasticsearch#39946
elastic/elasticsearch#45115
elastic/elasticsearch#42133

Logs showing the exception
---
Searching:
https://gist.github.com/wezell/ab8e3ebaf33a5f03c09de8f7c85c4920

Reindexing:
https://gist.github.com/wezell/06f25aa7f82aba26c3cb3b3be44c2985

### PR DESCRIPTION
**Pull request by wezell**
-- Description by @dsilvam 
ES relies on the httpclient lib and the later has an internal reactor for async events which once stopped makes the ES REST High Level Client non-usable. This leads to being unable to perform any requests against ES unless the server is restarted, even if the underlying cause/issue is solved. 

These changes detect the scenario when the reactor is stopped and then recreate the ES REST High Level Client.

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java [hunk 1]
```diff
@@ -1580,6 +1581,12 @@ SearchHits cachedIndexSearch(final SearchRequest searchRequest) {
         
     }
 
+    private void rebuildRestHighLevelClientIfNeeded(Exception e) {
+        if(e.getMessage().contains("reactor status: STOPPED")) {
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java:1585]: I would null check `e.getMessage()` before the contains.

Also, should we `if(e instanceof IllegalStateException)`?

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/RestHighLevelClientProvider.java [hunk 2]
```diff
@@ -79,5 +79,7 @@ private static boolean disableStringIntern() {
     
     public abstract RestHighLevelClient getClient();
 
+    public abstract void rebuildClient();
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/RestHighLevelClientProvider.java:82]: I would add a default impl for `rebuildClient()` that does nothing but log a message so customers who have overridden this class and have not seen issues will not have to rewrite their custom implementation to upgrade.

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java (hunk 1 above)
Comment (wezell) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java:1585]: maybe as a separate catch?

Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java]: Done

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/RestHighLevelClientProvider.java (hunk 2 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/util/RestHighLevelClientProvider.java]: Done

**Comment by dsilvam** [comment]
@wezell code-review done

**Review by freddyucv — COMMENTED** [review]
is it possible so any test here?

**Review by freddyucv — APPROVED** [review]

**Pull request merged**